xarray/core/dataarray.py

```diff
@@ -930,7 +930,10 @@ def copy(self, deep: bool = True, data: Any = None) -> "DataArray":
         """
         variable = self.variable.copy(deep=deep, data=data)
         coords = {k: v.copy(deep=deep) for k, v in self._coords.items()}
-        indexes = self._indexes
+        if self._indexes is None:
+            indexes = self._indexes
+        else:
+            indexes = {k: v.copy(deep=deep) for k, v in self._indexes.items()}
```

Comment:
I had thought that because indexes were immutable, we wouldn't need to deep copy, though the linked issue shows there's a flaw in that argument. 

This may have some performance implications. 

Do we need this change for this PR? We could merge the rest otherwise.

Reply:
Should deepcopy the `indexes` object rather than the `index` itself?

@toddrjen can you open a separate PR with this fix please?

Reply:
Yes, the change is required for this PR.  My tests depend on it.  I guess I could conceivably reorganize them to work without it, but that is quite a bit of work for a problem that is already fixed.

Reply:
I think this change _should_ be safe. It only does a deep copy of indexes if someone explicitly writes `array.copy(deep=True)`, in which I would guess this is desired behavior.

Ideally `pandas.Index` objects would have immutable values, but that isn't the case at the moment: https://github.com/pandas-dev/pandas/issues/33001

Reply:
@dcherian  As far as I can tell the `indexes` object is just a `dict`, which doesn't have a `deepcopy` option.  That is why I do it manually.